R/cube-transforms.R

```diff
@@ -126,24 +127,28 @@ setMethod("subtotalArray", "CrunchCube", function(x, headings = FALSE) {
 #' @aliases subtotalArray
 #' @export
 applyTransforms <- function(x,
-                            array = cubeToArray(showMissing(x)),
+                            array = cubeToArray(x),
                             transforms_list = transforms(x),
                             dims_list = dimensions(x),
                             useNA = x@useNA,
                             ...) {
     trans_indices <- which(vapply(transforms_list, Negate(is.null), logical(1)))
-    
+
     if (!missing(x)) {
         # we don't need to try and transform ca_items!
-        trans_indices <- trans_indices[getDimTypes(x)[trans_indices] != "ca_items"]   
+        trans_indices <- trans_indices[getDimTypes(x)[trans_indices] != "ca_items"]
     }
 
+    # evaluate which categories should be kept
+    keep_cats <- evalUseNA(array, dims_list, useNA = useNA)
+
     # Try to calculate the transforms for any dimension that has them, but
     # fail silently if they aren't calculable We use a for loop so that failing
     # one dimension doesn't break others. We could possibly use something like
     # abind::abind here and bind together the insertion vectors in array form,
     # but that would add a dependency
     errors <- list()
+    na_mask <- data.frame()
     for (d in trans_indices) {
```

Comment:
I'll see what I can do, in my looking at this before I found that we had to build up the `NA` mask inside the for loop because as we add insertions on each new dimension there are new cells that might need to be `NA` for the next (and subsequent) iterations.

A possibility that I investigated and rejected when drafting this was to allow all insertions to be used as inputs to other insertions, and then removing the ones at the end that don't make sense. This ended up being quite a bit more difficult and was a bit error prone.